### PR TITLE
Refactor Core. Integration of RuntimeData

### PR DESCRIPTION
**Pull request by junmendoza**
@lukechurch @ke-yu 
This will be one of many incremental changes in core componentization.
See the main task for details
http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-5540

[Source/AST]->[Codegen]->[Generate runtime data + exe]->[VM] 
(Runtime VM only consumes the exe and runtimecore)

1. Stub class for runtimecore and compilecore.  
2. Integrate runtimecore to the compiler and runtime.

You will notice that in most of these files, runtimecore is passed in together with core.
This looks redundant, but the point of this is to determine where the runtimecore should go in.
Succeeding submissions will slowly purge out the instances of core from the runtime.

Refactor submission are guaranteed to have no regressions unless otherwise stated.

Alternatively, I can setup a branch (CoreRefator) and just checkin all incremental changes in there. 
At the end of the task, merge in the entire CoreRefator branch to master

Let me know which approach is preferable

### CODE REVIEW

**Comment by junmendoza** [comment]
@lukechurch @RodRecker 
As discussed, ill make the incremental changes on this branch in this PR
This branch will be kept track of until the final merge to master.

**Comment by lukechurch** [comment]
Hey @junmendoza I think you misunderstood our email.

- We would like this to be a progress series of incremental PRs to master, not a final large merge.

**Comment by junmendoza** [comment]
@lukechurch yea thanks for the clarification. It would still be inevitable though that the first or first few merges to master would require 60+ files changes due to the references to core.
PR coming soon!

**Comment by junmendoza** [comment]
@lukechurch 

This initial PR integrates RuntimeData.
RuntimData is a datastructure containing the following:
1. Compiler generated - Read-only. Consumed by runtime VM
2. Compiler generated - R/W. Consumed by runtime VM
3. Runtime VM generated. R/W. Consumed by runtime VM

Currently, a few properties and methods from Core including trace related, have been moved to RuntimeData.
You will notice that it should be part of the DSExecutable and it is. It is currently accessed by the VM through exe.RuntimeData. This is integrated (already being used) at the VM and LiveRunner.

The next step will be to move the rest of the runtime properties from Core to RuntimeData.

File: src/Engine/ProtoCore/Core.cs [hunk 1]
```diff
@@ -1478,214 +1435,6 @@ private void ResetAll(Options options)
             cancellationPending = false;
         }
 
-        #region Trace Data Serialization Methods/Members
```
Comment (lukechurch) [inline src/Engine/ProtoCore/Core.cs]: Is this dead code that is being removed?

File: src/Engine/ProtoCore/DSASM/Executive.cs [hunk 2]
```diff
@@ -1051,59 +1048,28 @@ private bool GetSymbolIndexedIntoList(string symbol, out List<int> indexList)
...
 
         private FunctionCounter FindCounter(int funcIndex, int classScope, string name)
         {
-            foreach (FunctionCounter c in core.funcCounterTable)
+            foreach (FunctionCounter c in exe.RuntimeData.funcCounterTable)
             {
-
                 if (c.classScope == classScope && c.functionIndex == funcIndex)
                 {
-                    // Comment it out. Looks this foreach loop is dead code.
```
Comment (lukechurch) [inline src/Engine/ProtoCore/DSASM/Executive.cs]: In general this should be in a seperate PR

File: src/Engine/ProtoCore/RuntimeData.cs [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    /// 
+    /// The runtime VM is designed to run independently from the front-end (UI, compiler) 
+    /// and the only 2 properties it needs are the RuntimeData and the DSExecutable.
+    /// 
+    /// The RuntimeData will also contain properties that are populated at runtime and consumed at runtime.
+    /// </summary>
+    public class RuntimeData
+    {
+
+        #region COMPILER_GENERATED_READ_ONLY
```
Comment (lukechurch) [inline src/Engine/ProtoCore/RuntimeData.cs]: Can we mark them read only?

File: src/Engine/ProtoCore/Core.cs [hunk 4]
```diff
@@ -932,23 +930,20 @@ public enum ParseMode
...
     {
-        #region RUNTIME_PROPERTIES
 
         /// <summary>
-        ///  These are the list of symbols updated by the VM after an execution cycle
+        /// Properties in under COMPILER_GENERATED_TO_RUNTIME_DATA, are generated at compile time, and passed to RuntimeData/Exe
+        /// Only Core can initialize these
         /// </summary>
-        public HashSet<SymbolNode> UpdatedSymbols { get; private set; }
+        #region COMPILER_GENERATED_TO_RUNTIME_DATA
```
Comment (lukechurch) [inline src/Engine/ProtoCore/Core.cs]: Compile directives shouldn't have white space in front of them so they're easier to see

File: test/Engine/ProtoTest/Associative/RedefinitionExpression.cs [hunk 5]
```diff
@@ -1,106 +1,202 @@
-using System;
```
Comment (lukechurch) [inline test/Engine/ProtoTest/Associative/RedefinitionExpression.cs]: Something is borky in your setup I think? DO the difs really look like this?

File: test/Engine/ProtoTest/Associative/MicroFeatureTests.cs [hunk 6]
```diff
@@ -684,7 +684,7 @@ public void TestArrayOverIndexing01()
...
[Imperative]
{
    arr1 = {true, false};
    arr2 = {1, 2, 3};
    arr3 = {false, true};
    t = arr2[1][0];
}
";
             thisTest.RunScriptSource(code);
-            TestFrameWork.VerifyRuntimeWarning(ProtoCore.RuntimeData.WarningID.kOverIndexing);
```
Comment (lukechurch) [inline test/Engine/ProtoTest/Associative/MicroFeatureTests.cs]: False change, please do a discard local changes on this file

File: test/Engine/ProtoTest/EventTests/EventTest.cs [hunk 7]
```diff
@@ -1,220 +1,256 @@
-using System;
```
Comment (lukechurch) [inline test/Engine/ProtoTest/EventTests/EventTest.cs]: GitHub's diff algorithm is really really dumb at times.

Could you stick a note on the diff as to what has actualy changed here?

File: test/Engine/ProtoTestConsoleRunner/Program.cs [hunk 8]
```diff
@@ -28,6 +28,7 @@ static void Run(string filename, bool verbose)
             ProtoFFI.DLLFFIHandler.Register(ProtoFFI.FFILanguage.CSharp, new ProtoFFI.CSModuleHelper());
 
             ProtoScriptTestRunner runner = new ProtoScriptTestRunner();
+
             ExecutionMirror mirror = runner.LoadAndExecute(filename, core);
```
Comment (lukechurch) [inline test/Engine/ProtoTestConsoleRunner/Program.cs]: Discard local change

File: test/Engine/ProtoTestFx/DebugService.cs [hunk 9]
```diff
@@ -72,7 +72,9 @@ public bool RunScript(string dsPath)
                 core.Executives.Add(ProtoCore.Language.kImperative, new ProtoImperative.Executive(core));
                 core.Configurations.Add(Autodesk.DesignScript.Interfaces.ConfigurationKeys.GeometryFactory, GeometryFactoryName);
                 core.Configurations.Add(Autodesk.DesignScript.Interfaces.ConfigurationKeys.PersistentManager, PersistenceManagerName);
+                
                 ProtoScript.Runners.ProtoScriptTestRunner fsr = new ProtoScript.Runners.ProtoScriptTestRunner();
```
Comment (lukechurch) [inline test/Engine/ProtoTestFx/DebugService.cs]: Discard local change

File: test/Engine/ProtoTestFx/DebugTestFx.cs [hunk 10]
```diff
@@ -105,6 +104,7 @@ internal  static ProtoCore.Core DebugRunnerRunOnly(string code)
 
             runnerConfig = new ProtoScript.Config.RunConfiguration();
             runnerConfig.IsParrallel = false;
+
             fsr = new DebugRunner(core);
 
             DLLFFIHandler.Register(FFILanguage.CSharp, new CSModuleHelper());
```
Comment (lukechurch) [inline test/Engine/ProtoTestFx/DebugTestFx.cs]: Discard local change

File: test/Engine/ProtoTestFx/GeometryTestFrame.cs [hunk 11]
```diff
@@ -111,7 +111,6 @@ internal static double GetSelfDefinedTolerance()
             DLLFFIHandler.Register(FFILanguage.CSharp, new CSModuleHelper());
 
             //Run
```
Comment (lukechurch) [inline test/Engine/ProtoTestFx/GeometryTestFrame.cs]: Discard local change

File: src/Engine/ProtoCore/CompileData.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
+using System;
```
Comment (lukechurch) [inline src/Engine/ProtoCore/CompileData.cs]: This really belongs in a different PR

**Comment by lukechurch** [comment]
This is a good start - LGTM after nits are addressed.

It's close to the right break apart into PRs. it's usually worth:

- taking a quick glance over the PR before submitting and adjusting to remove non-changes,
- submit dead code elimination, that you're absolutely sure of, and doesn't require a conversation with anyone as a seperate TBR PR
- When the GitHub diff is impossible to read, just take a look and stick a comment on it saying what the actual change is

Thanks! Glad to see this moving fowards

File: src/Engine/ProtoCore/CompileData.cs (hunk 12 above)
Comment (junmendoza) [inline src/Engine/ProtoCore/CompileData.cs]: Removed for now

File: src/Engine/ProtoCore/Core.cs (hunk 4 above)
Comment (junmendoza) [inline src/Engine/ProtoCore/Core.cs]: Done

File: src/Engine/ProtoCore/RuntimeData.cs (hunk 3 above)
Comment (junmendoza) [inline src/Engine/ProtoCore/RuntimeData.cs]: Yes agree, will follow this up with all intended read-only in the next immediate PR

File: test/Engine/ProtoTest/Associative/RedefinitionExpression.cs (hunk 5 above)
Comment (junmendoza) [inline test/Engine/ProtoTest/Associative/RedefinitionExpression.cs]: After checking the file, this is a git diff issue. 
It probably detected a false change. 
There were no intended changes for this file

File: test/Engine/ProtoTest/EventTests/EventTest.cs (hunk 7 above)
Comment (junmendoza) [inline test/Engine/ProtoTest/EventTests/EventTest.cs]: As with the prev diff issues...this file has no intended change at all

File: src/Engine/ProtoCore/Core.cs [hunk 13]
```diff
@@ -2116,10 +1865,20 @@ public void GenerateExprExeInstructions(int blockScope)
             }
         }
 
+        private RuntimeData GenerateRuntimeData()
+        {
+            Validity.Assert(RuntimeData != null);
+            RuntimeData.FunctionTable = FunctionTable;
```
Comment (lukechurch) [inline src/Engine/ProtoCore/Core.cs]: This method mixes functional (returning the value) and state change (RuntimeData.FunctionTable = xxx) patterns.

This means that it's really surprising that it has both side effects and a return value. Lets pick one patten and stick to it.

File: src/Engine/ProtoCore/Executive.cs [hunk 14]
```diff
@@ -5,7 +5,7 @@ namespace ProtoCore
 {
 	public abstract class Executive
 	{
-        protected Core core; 
+        protected Core core;
```
Comment (lukechurch) [inline src/Engine/ProtoCore/Executive.cs]: ws

File: src/Engine/ProtoCore/Core.cs (hunk 13 above)
Comment (junmendoza) [inline src/Engine/ProtoCore/Core.cs]: Ok, the intention was to create a new runtimedata and populate it. will follow this ip in next PR to make sure no regression happen when this is merged
